prov/gni/src/gnix_buddy_allocator.c

```diff
@@ -0,0 +1,353 @@
+		│ BASE │ len  │ min │ max │ nlists │ LISTS │
+		└──────┴──────┴─────┴─────┴────────┴───────┘
+ * The LISTS pointer points to an array of dlist structures, each containing a
+ * head pointer to the begging of a free list.  Note that the first element of
+ * LISTS is a pointer to the head of the "min block size" free list, the second
+ * element is the head of the "min * 2 block size" free list and so on.
+
+ 		Node format as stored in a free block:
+		┌──────┬──────┬──────────────────────┐
+		│ NEXT │ PREV │ Remaining free bytes │
+		└──────┴──────┴──────────────────────┘
+ * Each NEXT and PREV pointer is stored in the first 16 bytes of the free block.
+ * This means that there is a hard limit of 16 bytes on the minimum block size.
+ *
+ *		Bitmap layout with a min block size of 16:
+ 		┌──┬──┬──┬──┬──┬──┬──┬──┬──┬──┬──┬──┬──┬──┬──┐
+		│16│16│16│16│..│32│32│32│32│..│64│64│64│64│..│
+		└──┴──┴──┴──┴──┴──┴──┴──┴──┴──┴──┴──┴──┴──┴──┘
+ * Each number above (16, 32, and 64) represents a block size that can
+ * be split. The first two "32" blocks are redundantly mapped by the the first
+ * 64 block, all blocks are redundant mappings of the next block size.
+ *
+ * When a block is split or allocated, its bit will be set in the bitmap.
+ * When a block is coalesced or free'd its bit will be reset in the bitmap.
+ * This improves the performance of coalescing blocks by being able to tell
+ * whether a given block's buddy block is allocated or split in O(c).
+ *
+ * TODO: dlist_insert_sorted for fragmentation reduction.
+ * TODO: Lock in __gnix_buddy_split and allow __gnix_buddy_find_block to run
+ * concurrently.
+ */
+
+#include "gnix_buddy_allocator.h"
+
+static inline int __gnix_buddy_create_lists(handle_t *alloc_handle)
+{
+	size_t i, offset = 0;
+
+	alloc_handle->nlists = (size_t) log2(alloc_handle->max /
+					     (double) MIN_BLOCK_SIZE) + 1;
+	alloc_handle->lists = calloc(1, sizeof(struct dlist_entry) *
+				     alloc_handle->nlists);
+
+	if (!alloc_handle->lists) {
+		GNIX_WARN(FI_LOG_EP_CTRL,
+			  "Could not create buddy allocator lists.\n");
+		return -FI_ENOMEM;
+	}
+
+	for (i = 0; i < alloc_handle->nlists; i++) {
+		dlist_init(alloc_handle->lists + i);
+	}
+
+	/* Insert free blocks of size max in sorted order into last list */
+	for (i = 0; i < alloc_handle->len / alloc_handle->max; i++) {
+		dlist_insert_tail(alloc_handle->base + offset,
+				  alloc_handle->lists +
+				  alloc_handle->nlists - 1);
+		offset += alloc_handle->max;
+	}
+
+	return FI_SUCCESS;
+}
+
+/**
+ * Split a block in list "j" until list "i" is reached.
+ */
+static inline void __gnix_buddy_split(handle_t *alloc_handle, size_t j, size_t i)
+{
+	void *tmp = NULL;
+
+	dlist_remove(tmp = alloc_handle->lists[j].next);
+
+	_gnix_set_bit(&alloc_handle->bitmap,
+		      BITMAP_INDEX(tmp, alloc_handle->base, MIN_BLOCK_SIZE,
+				   OFFSET(MIN_BLOCK_SIZE, j)));
+
+
+	/* Split the block until we reach list "i" */
+	for (j--; j > i; j--) {
```

Comment:
Okay, the bitmap is setup like this:

All the blocks that the buddy allocator allocates from the base block are of size X, where X = MBS * 2^Z, MBS is the minimum block size and Z is a non-negative integer.

The bitmap has 2 * (Len / MBS) bits and it's setup so that the first Len / MBS bits in the bitmap flag each block of size MBS as free or allocated.

Len is the number of bytes in the base block.
The base block is pointed to by void *base.

The first bit in the bitmap flags the first block of size MBS.  The first block of size MBS uses the address range:
base to (base + MBS - 1).

The second bit in the bitmap flags the second block of size MBS.  The second block of size MBS uses the address range:
(base + MBS) to (base + 2 * MBS - 1)

The third bit in the bitmap flags the third block of size MBS.  The third block of size MBS uses the address range:
(base + 2 * MBS) to (base + 3 * MBS - 1)

And so on until we reach the Len / MBS bit in the bitmap.

The second Len / MBS bits in the bitmap flag the remaining blocks of size X as free allocated, or split where X > MBS.

So, the first bit in the second Len / MBS bits in the bitmap flags the first block of size MBS * 2.  The first block of size MBS * 2 uses the address range:
base to (base + MBS * 2 - 1)

And so on until we reach the next block size.


The bitmap is used like this:

A bit is set to 1 when a block is allocated, or when the block is split into two smaller blocks.

A bit is reset to 0 when a block is freed, or when a free block is coalesced with another already free and unsplit block.

The bitmap is only read for coalescing blocks. When a block Y is freed we look at the bit in the bitmap for the buddy block of Y, if that bit is set then the buddy of Y is allocated, split, or both in which case we cannot coalesce Y with its buddy block.  However, if the bitmap bit for the buddy of Y is reset, then the buddy block of Y is free and not split, so we coalesce Y with the buddy of block of Y and continue to coalesce this new larger block with its buddy block until we reach the max block size or a buddy block that is allocated, split, or both.

I hope that makes sense, let me know if it doesn't.
> On Jan 8, 2016, at 11:04 AM, James Swaro <notifications@github.com> wrote:
> 
> In prov/gni/src/gnix_buddy_allocator.c:
> 
> > +/**
> > + * Split a block in list "j" until list "i" is reached.
> > + */
> > +static inline void __gnix_buddy_split(handle_t *alloc_handle, size_t j, size_t i)
> > +{
> > +	void *tmp = NULL;
> > +
> > +	dlist_remove(tmp = alloc_handle->lists[j].next);
> > +
> > +	_gnix_set_bit(&alloc_handle->bitmap,
> > +		      BITMAP_INDEX(tmp, alloc_handle->base, MIN_BLOCK_SIZE,
> > +				   OFFSET(MIN_BLOCK_SIZE, j)));
> > +
> > +
> > +	/* Split the block until we reach list "i" */
> > +	for (j--; j > i; j--) {
> I suppose I'm not worried about the space usage anymore, but I'm still not sure how it is that you use the bitmap and how you get from 0's and 1's to 16's and 32's .
> 
> —
> Reply to this email directly or view it on GitHub.
>